src/scheduled_tasks.rs

```diff
@@ -31,111 +31,138 @@ use lemmy_utils::{
   error::{LemmyError, LemmyResult},
   REQWEST_TIMEOUT,
 };
-use reqwest::blocking::Client;
-use std::{thread, time::Duration};
+use reqwest::Client;
+use std::time::Duration;
 use tracing::{error, info, warn};
 
 /// Schedules various cleanup tasks for lemmy in a background thread
-pub fn setup(
+pub async fn setup(
   db_url: String,
   user_agent: String,
   context_1: LemmyContext,
 ) -> Result<(), LemmyError> {
   // Setup the connections
-  let mut scheduler = Scheduler::new();
+  let mut scheduler = AsyncScheduler::new();
 
   startup_jobs(&db_url);
 
   // Update active counts every hour
   let url = db_url.clone();
   scheduler.every(CTimeUnits::hour(1)).run(move || {
-    PgConnection::establish(&url)
-      .map(|mut conn| {
-        active_counts(&mut conn);
-        update_banned_when_expired(&mut conn);
-      })
-      .map_err(|e| {
-        error!("Failed to establish db connection for active counts update: {e}");
-      })
-      .ok();
+    let url = url.clone();
+    async move {
+      PgConnection::establish(&url)
+        .map(|mut conn| {
+          active_counts(&mut conn);
+          update_banned_when_expired(&mut conn);
+        })
+        .map_err(|e| {
+          error!("Failed to establish db connection for active counts update: {e}");
+        })
+        .ok();
+    }
   });
 
-  // Update hot ranks every 15 minutes
   let url = db_url.clone();
-  scheduler.every(CTimeUnits::minutes(15)).run(move || {
-    PgConnection::establish(&url)
-      .map(|mut conn| {
-        update_hot_ranks(&mut conn);
-      })
-      .map_err(|e| {
-        error!("Failed to establish db connection for hot ranks update: {e}");
-      })
-      .ok();
+  // Update hot ranks every 15 minutes
+  scheduler.every(CTimeUnits::minutes(10)).run(move || {
+    let url = url.clone();
+    async move {
+      PgConnection::establish(&url)
+        .map(|mut conn| {
+          update_hot_ranks(&mut conn);
+        })
+        .map_err(|e| {
+          error!("Failed to establish db connection for hot ranks update: {e}");
+        })
+        .ok();
+    }
   });
 
   // Delete any captcha answers older than ten minutes, every ten minutes
   let url = db_url.clone();
   scheduler.every(CTimeUnits::minutes(10)).run(move || {
-    PgConnection::establish(&url)
-      .map(|mut conn| {
-        delete_expired_captcha_answers(&mut conn);
-      })
-      .map_err(|e| {
-        error!("Failed to establish db connection for captcha cleanup: {e}");
-      })
-      .ok();
+    let url = url.clone();
+    async move {
```

Comment:
Is it really necessary to have move and then async move inside?

Reply:
It's because of `url` . Need to get that inside `AsyncJob.run` and then into the async block, that I think we need because there's no async version of the diesel `PgConnection`. Couldn't think of a better way to do this 🤔